src/dataflow/src/render/reduce.rs

```diff
@@ -195,185 +624,147 @@ where
-                                for typ in aggregate_types.iter() {
-                                    match typ {
-                                        ReductionType::Accumulable => {
-                                            row_packer.push(accumulable.next().unwrap())
-                                        }
-                                        ReductionType::Hierarchical => {
-                                            row_packer.push(hierarchical.next().unwrap())
-                                        }
-                                        ReductionType::Basic => {
-                                            row_packer.push(basic.next().unwrap())
-                                        }
-                                    }
-                                }
-                                output.push((row_packer.finish_and_reuse(), 1));
-                            }
-                        })
-                };
-                let index = (0..group_key.len()).collect::<Vec<_>>();
-                self.set_local_columns(
-                    relation_expr,
-                    &index[..],
-                    (arrangement, err_input.arrange()),
-                );
+                }
+                output.push((row_packer.finish_and_reuse(), 1));
             }
+        })
+}
+
+fn build_distinct<G>(collection: Collection<G, (Row, Row)>) -> Arrangement<G, Row>
+where
+    G: Scope,
+    G::Timestamp: Lattice,
+{
+    collection.reduce_abelian::<_, OrdValSpine<_, _, _, _>>("DistinctBy", {
+        |key, _input, output| {
+            output.push((key.clone(), 1));
         }
-    }
+    })
 }
 
-fn build_basic<G>(
-    collection: Collection<G, (Row, Row)>,
+/// Build the dataflow to compute and arrange multiple non-accumulable,
+/// non-hierarchical aggregations on `input`.
+///
+/// This function assumes that we are explicitly rendering multiple basic aggregations.
+/// For each aggregate, we render a different reduce operator, and then fuse
+/// results together into a final arrangement that presents all the results
+/// in the order specified by `aggrs`. `prepend_keys` is true if the arrangement
+/// produced by this function needs to be reused by other views.
+fn build_basic_aggregates<G>(
+    input: Collection<G, (Row, Row)>,
     aggrs: Vec<(usize, AggregateExpr)>,
     prepend_key: bool,
 ) -> Arrangement<G, Row>
 where
     G: Scope,
     G::Timestamp: Lattice,
 {
-    if aggrs.len() == 1 {
-        // If we just have a single basic aggregation we can just arrange that without doing
-        // any fusion.
-        return build_basic_aggregate(collection, aggrs[0].0, &aggrs[0].1, prepend_key);
-    }
-
-    // Otherwise we need to render each individually and stitch them together.
+    // We are only using this function to render multiple basic aggregates and
+    // stitch them together
+    assert!(aggrs.len() > 1);
```

Comment:
This stands out as a risky behavioral change. Is the code hard to write in this case, and we should log an error, or is it important to take down Materialize at this point? I think philosophy elsewhere is to panic on unrecoverable errors, which this doesn't seem to be.

Reply:
codes not hard to write. i just wanted a stronger check that this code was only being called if we had multiple basic aggregates. downgraded this to a log